mayastor/src/bdev/nexus/nexus_metadata.rs

```diff
@@ -0,0 +1,620 @@
+        let alignment = bdev.alignment();
+
+        let self_lba = metadata.header.self_lba;
+        let mut start = metadata.header.data_start;
+
+        for entry in &mut metadata.index {
+            if entry.data_start > start {
+                let blocks = entry.data_end - entry.data_start;
+                let mut buf = DmaBuf::new(
+                    ((blocks + 1) * block_size) as usize,
+                    alignment,
+                )
+                .context(ObjectAlloc {})?;
+                self.read_at(
+                    (self_lba + entry.data_start) * block_size,
+                    &mut buf,
+                )
+                .await
+                .context(ObjectRead {})?;
+                self.write_at((self_lba + start) * block_size, &buf)
+                    .await
+                    .context(ObjectWrite {})?;
+                entry.data_start = start;
+                entry.data_end = start + blocks;
+            }
+            start = entry.data_end + 1;
+        }
+
+        Ok(())
+    }
+
+    // Update checksums and write out MetaData header + index to disk.
+    pub async fn sync_metadata(
+        &mut self,
+        metadata: &mut NexusMetaData,
+    ) -> Result<(), MetaDataError> {
+        metadata.header.index_checksum =
+            MetaDataIndexEntry::checksum(&metadata.index);
+        metadata.header.checksum();
+        self.write_index(&metadata).await
+    }
+
+    // Create a new header + index on "MetaData" partition.
+    pub async fn create_metadata(
+        &mut self,
+    ) -> Result<NexusMetaData, MetaDataError> {
+        let (bdev, _desc) = self.get_dev()?;
+
+        if let Some(partition) = self.probe_label().await?.partitions.get(0) {
+            if partition.ent_type
+                == GptGuid::from_str("27663382-e5e6-11e9-81b4-ca5ca5ca5ca5")
+                    .unwrap()
+                && partition.ent_name.name == "MayaMeta"
+            {
+                let mut metadata = NexusMetaData {
+                    header: MetaDataHeader::new(bdev.block_len(), &partition),
+                    index: Vec::new(),
+                };
+                self.sync_metadata(&mut metadata).await?;
+                return Ok(metadata);
+            }
+        }
+
+        Err(MetaDataError::MissingPartition {})
+    }
+
+    // Retrieve header + index from "MetaData" partition.
+    pub async fn get_metadata(&self) -> Result<NexusMetaData, MetaDataError> {
+        if let Some(partition) = self.probe_label().await?.partitions.get(0) {
+            if partition.ent_type
+                == GptGuid::from_str("27663382-e5e6-11e9-81b4-ca5ca5ca5ca5")
+                    .unwrap()
+                && partition.ent_name.name == "MayaMeta"
+            {
+                return Ok(self.probe_index(partition.ent_start).await?);
+            }
+        }
+
+        Err(MetaDataError::MissingPartition {})
+    }
+
+    // Retrieve specified config object from "MetaData" partition.
```

Comment:
I assume get the object at the at index N? the comment is not really clear ifso can we rename n to index?